File: TobiiExampleScript/Tobii_helpers/t2t/examples/perl/calibration.pl
#!/usr/bin/env perl -w
##
# Luca Filippin - July 2010 - luca.filippin@gmail.com                                                
# Copyright 2010 Language, Cognition and  Development Laboratory, Sissa, Trieste. All rights reserved.
# 
# PLEASE HAVE A LOOK TO THE SWIG DOCUMENTATION FOR A DEEPER UNDERSTANDING OF THIS CODE
# NOTE: Swig wrapper to t2t makes use of shadow classes and swig helpers (see t2tsw.i)
##

#use strict;
use File::Basename;
use Getopt::Long;
use t2tHelpers;
use Time::HiRes;
use Tk::CursorControl;
use Tk::JPEG;
use Tk::PNG;
use Tk;

GetOptions(	"help|h"=>\$opt_help,
			"version"=>\$opt_version,
			"log-file|l:s"=>\$opt_log_file,
   			"port|p:i"=>\$opt_tobii_port,
            "calibration-file|o:s"=>\$opt_calibration_file,
            "load-calibration|d"=>\$opt_calibration_load,
            "start-delay|s:f"=>\$opt_start_delay,
            "output-file|r:s"=>\$opt_output_file);
            
($command) = fileparse($0);

$version=<<VERSION;
+-------------------------------------------------------------------------------------+
|  calibration.pl v1.0                                                                |
|  This is a sample application which performs the calibration of a Tobii ET.         |
|-------------------------------------------------------------------------------------|
|  Luca Filippin - July 2010 - luca.filippin\@gmail.com                                |                                             
|  Copyright 2010 Language, Cognition and  Development Laboratory, Sissa, Trieste     |
+-------------------------------------------------------------------------------------+
VERSION

$usage=<<USAGE;

Usage: $command [options] <TET ip address> <input file>

       <input file> lines fmt: 'x[0,1] y[0,1] height(0,1] width(0,1] delay(ms) picture_file_name', field separator: tab

Calibrate a Tobii Eyes Tracker

Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit
  -l TOBII_LOG, --log-file=TOBII_LOG
                        log file name
  -p TOBII_PORT, --port=TOBII_PORT
                        TET server listening port
  -o CALIBRATION_FILE, --calibration-file=CALIBRATION_FILE
                        Name of the file which will store the calibration
                        results
  -d, --load-calibration
                        Use existing calibration: just print results
  -s START_DELAY, --start-delay=START_DELAY
                        Start calibration after a specific delay (secs)
  -r OUTPUT_FILE, --output-file=OUTPUT_FILE
                        Name of the file which will store the collected data
USAGE

if ($opt_version) {
	print "$version";
	exit 0;
}

if (@ARGV < 2 or $opt_help) {
  	print "$usage";
  	exit 0;
}

if (!defined($opt_tobii_port)) { $opt_tobii_port=4455; }
if (!defined($opt_start_delay)) { $opt_start_delay=0; }

$opt_tobii_ip = $ARGV[0];
$opt_input_file = $ARGV[1];

sub banner() {
	print "$version\n";
}

sub parseInputFile {
	my $filename = $_[0];
	
	my %data;
	$data{size}  = 0; 
	@$data{pict} = ();
	
	eval {
		open F, '<', $filename or die "Failed to OPEN the data file $filename";
		
		foreach (<F>) {
			if (/^(\d*\.?\d+)\t(\d*\.?\d+)\t(\d*\.?\d+)\t(\d*\.?\d+)\t(\d+)\t(.*)/) {
				
			   	die "Data line is in bad format: $_" if ($1<0 || $1>1 || $2<0 || $2>1 || $3<=0 || $3>1 || $4<=0 || $4>1 || $5<0);
			   	die "File is not existing or is not a readablefile: $6" unless ( -e $6 && -f $6 && -r $6); 
			   
			    my $rec = { 
			    	x => $1,
					y => $2,
					height => $3,
					width => $4,
					delay => $5,
					fname => $6,
				};
				
				$data{size}++;
				push @{$data{pict}}, $rec;
				next;
			}
			elsif (/^#*\s*$/) {
				next;
			}
			close F;
			die "Unexpected line: $_";
		}
		close F;
	} or do {
		die "++ Error reading input file\n   $@";
	};
	
	return \%data;
}

sub writeOutputFile {
	my ($filename, $data) = @_;
	my $ok = 1;
	
	eval {
		open F, '>', $filename or die $!;
		printf F  "%s\n%s", $data->compact_header(), $data->str(1);
		close F;
	} or do {
		$ok = 0;
	};
	return $ok;
}

sub get_image {
	my ($canvas, $w_width, $w_height, $pict) = @_;
	
	my $p1 = $canvas->Photo(-file => $pict->{fname});
	my $r_w = ($w_width * $pict->{width})/$p1->width;
	my $r_h = ($w_height * $pict->{height})/$p1->height;
	my $subs_x = $r_w < 1 ? 1/$r_w : 1;
	my $subs_y = $r_h < 1 ? 1/$r_h : 1;
	my $zoom_x = $r_w > 1 ? $r_w : 1;
	my $zoom_y = $r_h > 1 ? $r_h : 1;
	
	if ($subs_x > 1 || $subs_y > 1) {
		my $p2 = $canvas->Photo();
		$p2->copy($p1, -shrink, -subsample => $subs_x, $subs_y);
		$p1 = $p2;
	}
	if ($zoom_x > 1 || $zoom_y > 1) {
		my $p2 = $canvas->Photo();
		$p2->copy($p1, -zoom => $zoom_x, $zoom_y);
		$p1 = $p2;
	}
	return $p1;
}

sub calibration {
	my ($canvas, $data, $i, $to_delete) = @_;
	my $main_w = $canvas->MainWindow;
	my $w = $main_w->screenwidth();
	my $h = $main_w->MainWindow->screenheight();
	
	if ($i < 0) {
		$main_w->fontCreate('big', -family=>'courier', -weight=>'bold', -size=>int(-18*18/12));
		my $msg = $canvas->createText($w/2, $h/2, -anchor =>'center', -font =>'big', -text => "Please focus on the appearing pictures...", -fill => "green");
		$canvas->update;
		$canvas->after(2000, \&calibration, $canvas, $data, $i+1, $msg);
		return;
	}
	if (defined $to_delete) {
		if ($i > 0) { 
			my $c = new t2tsw::t2tCmd();
        	$c->{cmd} = "DREW_POINT";
        	t2tHelpers::cmdEx($c);
		}
		$canvas->delete($to_delete);
		$canvas->update;
	}
	if ($i < $data->{size}) {
		my $p = $data->{pict}[$i];
		my $img = $canvas->create('image', $p->{x}*$w, $p->{y}*$h, -anchor =>'center', -image => $p->{image});
		$canvas->update;
		
		my $c = new t2tsw::t2tCmd();
        $c->{cmd} = "ADD_CALIBRATION_POINT";
        t2tHelpers::cmdEx($c);
		
		$canvas->after($data->{pict}[$i]->{delay}, \&calibration, $canvas, $data, $i+1, $img);
	} else {
		$main_w->destroy;
	}
}

eval {
	$disconnect = 0;
	banner();
	t2tsw::t2tOutputFileName($opt_log_file, "w") if defined $opt_log_file;
	
	if (! $opt_calibration_load) {
		$data = parseInputFile($opt_input_file);
		t2tHelpers::bailOut("Calibration needs at least 2 points\nExiting\n", 0) unless $data->{size} >= 2;
		
		# Tk initialisation
		$window = MainWindow->new;
		$window->configure(-title => "Tobii Eye Tracker Calibration");
		$window->CursorControl->hide($window);
		$w_height = $window->screenheight();
		$w_width = $window->screenwidth();
		$window->configure( -width=>$w_width, -height=>$w_height );
		$window->resizable( 0, 0 ); # not resizable in any direction
		$canvas = $window->Canvas(-background=>"black", -width=>$w_width, -height=>$w_height, -borderwidth => 0);	
		$canvas->pack(-expand => 1, -fill => 'both');
		
		$data->{images} = [];
		for ($i = 0; $i < $data->{size}; $i++) {
		    $data->{pict}[$i]->{image} = get_image($canvas, $w_width, $w_height, $data->{pict}[$i]);
		}
    }
    
	# Connect to the tobii
    print "Connecting to the TET server @".$opt_tobii_ip.":".$opt_tobii_port."...\n";
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "CONNECT";
    $c->{prm}->{connect}->{ip_address} = $opt_tobii_ip;
    $c->{prm}->{connect}->{port} = $opt_tobii_port;
    t2tHelpers::cmdEx($c);
	t2tHelpers::checkStatus('connect', 1, 5);
	t2tHelpers::bailOut("Connection failed\nExiting\n", 0) unless t2tHelpers::checkStatus('connected', 1, 5);
	$disconnect = 1;
    
    if ($opt_calibration_load) {
    	$c = new t2tsw::t2tCmd();
    	$c->{cmd} = "START_CALIBRATION";
    	$c->{prm}->{start_calibration}->{load_from_file} = 1;
    	$c->{prm}->{start_calibration}->{cmatrix}->{cols} = 0;
    	$c->{prm}->{start_calibration}->{cmatrix}->{rows} = 0;
    	$c->{prm}->{start_calibration}->{cmatrix}->{vals} = undef;
    	$c->{prm}->{start_calibration}->{fname} = $opt_calibration_file;
    	t2tHelpers::cmdEx($c);
    	
    	t2tHelpers::checkStatus('calibrating', 1, 5);
		t2tHelpers::bailOut("Load calibration failed\nExiting\n", 1) unless t2tHelpers::checkStatus('calibstarted', 1, 5);
		t2tHelpers::bailOut("Load calibration failed\nExiting\n", 1) unless t2tHelpers::checkStatus('calibstarted', 0, 5);
    } else {
    	if (defined $opt_start_delay && $opt_start_delay > 0) {
    		 Time::HiRes::usleep $opt_start_delay*1000;
    	} else {
    		t2tHelpers::rawInput("\n***** PRESS ANY KEY TO GET CALIBRATION DATA *****\n");
    	}
    	
    	# A tracking phase seems to be necessary before starting a calibration, or it will fails very often.
    	# There's no explanation or documentation about, but it works like that.
    	$c = new t2tsw::t2tCmd();
		$c->{cmd} = "START_TRACKING";
		t2tHelpers::cmdEx($c);
		t2tHelpers::checkStatus('running', 1, 5);
		t2tHelpers::bailOut("Start tracking failed\nExiting\n", 1) unless t2tHelpers::checkStatus('runstarted', 1, 5);
    	Time::HiRes::sleep 2;
    	$c = new t2tsw::t2tCmd();
		$c->{cmd} = "STOP_TRACKING";
		t2tHelpers::cmdEx($c);
		t2tHelpers::checkStatus('stop', 1, 5);
		t2tHelpers::bailOut("Stop tracking failed\nExiting\n", 0) unless t2tHelpers::checkStatus('running', 0, 5);
    	
    	$c = new t2tsw::t2tCmd();
    	$c->{cmd} = "START_CALIBRATION";
    	$c->{prm}->{start_calibration}->{clear_previous} = 1;
        $c->{prm}->{start_calibration}->{samples_per_point} = 20; 
    	$c->{prm}->{start_calibration}->{load_from_file} = 0;
    	$c->{prm}->{start_calibration}->{cmatrix}->{cols} = 2;
    	$c->{prm}->{start_calibration}->{cmatrix}->{rows} = $data->{size};
    	$c->{prm}->{start_calibration}->{cmatrix}->{vals} = $v = new t2tsw::doubleArrayC(2*$data->{size});
    	$c->{prm}->{start_calibration}->{fname} = $opt_calibration_file;
    	for ($i = 0; $i < $data->{size}; $i++) {
    		t2tsw::doubleArrayC::setitem($v, $i*2, $data->{pict}[$i]->{x});
    		t2tsw::doubleArrayC::setitem($v, $i*2+1, $data->{pict}[$i]->{y}); 
        }
        t2tHelpers::cmdEx($c);
        t2tHelpers::checkStatus('calibrating', 1, 5);
		t2tHelpers::bailOut("Start calibration failed\nExiting\n", 1) unless t2tHelpers::checkStatus('calibstarted', 1, 5);
		
		# Start the the displaying procedure...
		calibration($canvas, $data, -1);
		MainLoop;
		
		t2tHelpers::bailOut("End calibration delayed\nExiting\n", 1) unless t2tHelpers::checkStatus('calibend', 1, 30);
    }
    
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "CALIBRATION_ANALYSIS";
    $calib_an = t2tHelpers::cmdEx($c);
    
    if (defined($calib_an->{calibration_analysis})) {
        $print_data = (!defined($opt_output_file));
        if  (! $print_data) {
            if (! writeOutputFile($opt_output_file, $calib_an->{calibration_analysis})) {
                print "Error: can't save data file! Print to screen...";
                $print_data = 1;
            }
		}
        print "\nCalibration analisys data:\n\n".$calib_an->{calibration_analysis}->compact_header()."\n".$calib_an->{calibration_analysis}->str(1) if $print_data;
    } else {
        print "Calibration failed: no data\n";
    }
    
    $disconnect = 0;
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "DISCONNECT";
    t2tHelpers::cmdEx($c, 0);
    t2tHelpers::checkStatus('disconnect', 1, 5);
	t2tHelpers::bailOut("Disconnection failed\nExiting\n", 0) unless t2tHelpers::checkStatus('connected', 0, 5);
	
    $c = new t2tsw::t2tCmd();
    $c->{cmd} = "CLEANUP";
    t2tHelpers::cmdEx($c);
    printf "End.\n";
    
} or do {
	my $msg = "--------------- UNEXPECTED ERROR ---------------\n$@";
    $msg .= "------------------------------------------------\n";
    $msg .= "The application has been forcely closed!\n";
    t2tHelpers::bailOut($msg, $disconnect);
}

File: TobiiExampleScript/Tobii_helpers/t2t/examples/python/calibration.py
#!/usr/bin/env python

"""
# Luca Filippin - July 2010 - luca.filippin@gmail.com                                                
# Copyright 2010 Language, Cognition and  Development Laboratory, Sissa, Trieste. All rights reserved.
# 
# PLEASE HAVE A LOOK TO THE SWIG DOCUMENTATION FOR A DEEPER UNDERSTANDING OF THIS CODE
# NOTE: Swig wrapper to t2t makes use of shadow classes and swig helpers (see t2tsw.i)
"""

from argparse import ArgumentParser
from argparse import RawTextHelpFormatter
from t2tsw import *
from t2tHelpers import *
import sys
import os
import time
import re
import pygame

def banner():
    print ""
    print "+-------------------------------------------------------------------------------------+"
    print "|  This is a sample application which performs the calibration of a Tobii ET.         |"
    print "|-------------------------------------------------------------------------------------|"
    print "|  Luca Filippin - July 2010 - luca.filippin@gmail.com                                |"                                                
    print "|  Copyright 2010 Language, Cognition and  Development Laboratory, Sissa, Trieste     |"
    print "+-------------------------------------------------------------------------------------+"
    print ""

def parseOptions():
    parser = ArgumentParser(description='Calibrate a Tobii Eyes Tracker', formatter_class=RawTextHelpFormatter)
    parser.add_argument('tobii_ip', help='tobii ip address')
    parser.add_argument('input_file', help='calibration points file.\nfile line format: "x[0,1] y[0,1] height(0,1] width(0,1] delay(ms) picture_file_name"\nfield separator: tab')
    parser.add_argument('--version', action = 'version', version="%(prog)s v1.0")
    parser.add_argument("-l", "--log-file", dest="tobii_log", default="", help="log file name")
    parser.add_argument("-p", "--port", dest="tobii_port", type=int, default=4455, help="TET server listening port")
    parser.add_argument("-o", "--calibration-file", dest="calibration_file", default=None, help="Name of the file which will store the calibration results")
    parser.add_argument("-d", "--load-calibration", dest="calibration_load", action="store_true", help="Use existing calibration: just print results")
    parser.add_argument("-s", "--start-delay", dest="start_delay", type = float, default=0, help="Start calibration after a specific delay (secs)")
    parser.add_argument("-r", "--output-file", dest="output_file", default="", help="Name of the file which will store the collected data")
    
    try:
        o = parser.parse_args()
        if o.start_delay < 0:
            raise Exception("Bad delay value '%(v)s': should be > 0" %{ 'v' : o.start_delay })
        if o.tobii_port < 0:
            raise Exception("Bad port value '%(v)s': should be > 0" %{ 'v' : o.tobii_port })
    except Exception, e:
        raise Exception("++ Argument error\n" + e.__str__())
    
    return o

class calib_point_data(object):
    pass
    
def parseInputFile(filename):
    ignore_line_re = re.compile("^#*\s*$")
    data_line_re = re.compile("^(\d*\.?\d+)\t(\d*\.?\d+)\t(\d*\.?\d+)\t(\d*\.?\d+)\t(\d+)\t(.*)") # x y height width delay fname
    data = calib_point_data()
    data.size = 0
    data.x = []
    data.y = []
    data.height = []
    data.width = []
    data.delays = []
    data.pictfn = []

    try:
        F = None
        F = open(filename, 'r')
        data_lines = F.read().split('\n')
        
        for i in xrange(0, len(data_lines)):
            if re.match(ignore_line_re, data_lines[i]):
                continue
            m = re.match(data_line_re, data_lines[i])
            if m <> None:
                x = float(m.group(1))
                y = float(m.group(2))
                h = float(m.group(3))
                w = float(m.group(4))
                t = int(m.group(5))
                f = m.group(6)
                
                if (x < 0 or x > 1) or (y < 0 or y> 1) or (h <= 0 or h> 1) or (w <= 0 or w> 1)  or (t < 0):
                    raise Exception("  Data line is in bad format:\n" + data_lines[i])
                if not os.path.isfile(f):
                    raise Exception("  File is not existing or is not a file:\n" + f)
            else:
                raise Exception("  Unexpected line:\n" + data_lines[i])
                
            data.x.append(x)
            data.y.append(y)
            data.height.append(h)
            data.width.append(w)
            data.delays.append(t)
            data.pictfn.append(f)
         
        data.size = len(data.x)
        
    except Exception, e:
        raise Exception("++ Error reading input file\n" + e.__str__())
    finally:
        if F != None: 
            F.close()
    return data
    
def writeOutputFile(filename, data):
    ok = True
    try:
        F = None
        F = open(filename, 'w')
        F.write(t2tCalibrationData._compact_header + "\n" + data.__str__(compact = True)) 
    except Exception, e:
        ok = False
    finally:
        if F != None: 
            F.close()  
    return ok    

try:

    banner()
    tobii_connected = False
    o = parseOptions() 
    if o.tobii_log <> "": 
        t2tOutputFileName(o.tobii_log, "w")
    
    if not o.calibration_load:
        data = parseInputFile(o.input_file)
        if data.size < 2:
            bailOut(msg = "Calibration needs at least 2 points\nExiting", disconnect = False)
        
        # Initialise pygame and load the pictures in memory
        pygame.init()
        pygame.mouse.set_visible(0)
        pygame.display.set_caption("Tobii Eye Tracker Calibration")
        screen = pygame.display.set_mode()
        background = pygame.Surface(screen.get_size()).convert()
        bg_size = background.get_size()
        background.fill((0, 0, 0))
        data.pictures = []
        for i in xrange(0, data.size):
            p = pygame.image.load(data.pictfn[i]).convert()
            h, w = data.height[i], data.width[i]
            data.pictures.append(pygame.transform.scale(p, (int(bg_size[1]*w), int(bg_size[0]*h))))
    
    # Connect to the tobii
    print "Connecting to the TET server @" + o.tobii_ip + ":" + '%(p)d' %{ 'p':o.tobii_port } + "..."
    c = t2tCmd()
    c.cmd = "CONNECT"
    c.prm.connect.ip_address = o.tobii_ip
    c.prm.connect.port = o.tobii_port
    cmdEx(c)
    checkStatus("connect", 1, timeout = 5)
    if not checkStatus("connected", 1, timeout = 5): 
        bailOut(msg = "Connection failed\nExiting", disconnect = False)
    tobii_connected = True
    
    if o.calibration_load:        
        c = t2tCmd()
        c.cmd = "START_CALIBRATION"
        c.prm.start_calibration.load_from_file = 1
        c.prm.start_calibration.cmatrix.cols = 0
        c.prm.start_calibration.fname = o.calibration_file
        c.prm.start_calibration.cmatrix.rows = 0
        c.prm.start_calibration.cmatrix.vals = None
        cmdEx(c)
        
        checkStatus("calibrating", 1, timeout = 5)
        if not checkStatus("calibstarted", 1, timeout = 5): 
            bailOut(msg = "Load calibration failed\nExiting", disconnect = True)
        
        # there's no calibend flag set for loading
        if not checkStatus("calibrating", 0, timeout = 20): 
            bailOut(msg = "Load calibration failed\nExiting", disconnect = True)
            
    else:
        if o.start_delay == 0:
            raw_input("\n***** PRESS <enter> TO START CALIBRATION *****\n")
        else:
            time.sleep(o.start_delay)
        
        # A tracking phase seems to be necessary before starting a calibration, or it will fails very often.
    	# There's no explanation or documentation about, but it works like that.
    	c = t2tCmd()
        c.cmd = "START_TRACKING"
        cmdEx(c)
        checkStatus("running", 1, timeout = 5)
        if not checkStatus("runstarted", 1, timeout = 5):    
            bailOut(msg = "Start tracking failed\nExiting", disconnect = True)
    	time.sleep(2)
    	c = t2tCmd()
        c.cmd = "STOP_TRACKING"
        cmdEx(c)
        checkStatus("stop", 1, timeout = 1)
        if not checkStatus("running", 0, timeout = 5): 
            bailOut(msg = "Stop tracking failed\nExiting", disconnect = True)
        
        # Start now the calibration phase
        c = t2tCmd()
        c.cmd = "START_CALIBRATION"
        c.prm.start_calibration.clear_previous = 1
        c.prm.start_calibration.samples_per_point = 20 
        c.prm.start_calibration.load_from_file = 0
        c.prm.start_calibration.fname = o.calibration_file
        c.prm.start_calibration.cmatrix.cols = 2
        c.prm.start_calibration.cmatrix.rows = data.size
        vals = doubleArrayC(2*data.size)
        c.prm.start_calibration.cmatrix.vals = vals
        for i in xrange(0, data.size):
            vals[i*2] = data.x[i]
            vals[i*2+1] = data.y[i] 
        cmdEx(c)
        
        checkStatus("calibrating", 1, timeout = 5)
        if not checkStatus("calibstarted", 1, timeout = 5): 
            bailOut(msg = "Start calibration failed\nExiting", disconnect = True)
        
        if pygame.font:
            text = pygame.font.Font(None, 36).render("Please focus on the appearing pictures...", 1, (10, 10, 10))
            textpos = text.get_rect(centerx=background.get_width()/2)
            background.blit(text, textpos)
            pygame.display.flip()
            pygame.time.delay(2000)
            screen.blit(text, textpos, special_flags = pygame.BLEND_SUB)
            pygame.display.flip()
            
        # Do not check status while adding/drawing points 
        for i in xrange(0, data.size):
            x, y, h, w = data.x[i], data.y[i], data.height[i], data.width[i]
            screen.blit(data.pictures[i], ((x - w/2) * bg_size[0], (y - h/2) * bg_size[1]))
            pygame.display.flip()
            c = t2tCmd()
            c.cmd = "ADD_CALIBRATION_POINT"
            cmdEx(c)
            pygame.time.delay(data.delays[i])
            c = t2tCmd()
            c.cmd = "DREW_POINT"
            cmdEx(c)
            screen.blit(data.pictures[i], ((x - w/2)*bg_size[0], (y - h/2) * bg_size[1]), special_flags = pygame.BLEND_SUB)
            pygame.display.flip()  
        
        if pygame.font:
            text = pygame.font.Font(None, 36).render("The calibration has been completed. Thanks!", 1, (10, 10, 10))
            textpos = text.get_rect(centerx=background.get_width()/2)
            background.blit(text, textpos)
            pygame.display.flip()
            pygame.time.delay(2000)
            screen.blit(text, textpos, special_flags = pygame.BLEND_SUB)
            pygame.display.flip()
   
        pygame.display.quit()
        
        if not checkStatus("calibend", 1, timeout = 30): 
            bailOut(msg = "End calibration delayed... \nExiting", disconnect = True)
    
    c = t2tCmd()
    c.cmd = "CALIBRATION_ANALYSIS"  
    calib_an = cmdEx(c).calibration_analysis
    if calib_an <> None:
        print_data = (o.output_file == "")
        if not print_data:
            if not writeOutputFile(o.output_file, calib_an):
                print "Error: can't save data file! Print to screen..."
                print_data = True
        if print_data:
            print "\nCalibration analisys data:\n\n" + t2tCalibrationData._compact_header + "\n" + calib_an.__str__(compact = True)
    else:
        print "Calibration failed: no data"
    
    tobii_connected = False
    c = t2tCmd()
    c.cmd = "DISCONNECT"
    cmdEx(c, disconnect = False)
    checkStatus("disconnect", 1, timeout = 5)
    if not checkStatus("connected", 0, timeout = 5): 
        bailOut(msg = "Disconnect failed\nExiting", disconnect = False)
    
    c = t2tCmd()
    c.cmd = "CLEANUP"
    cmdEx(c)
    
    print "End.\n"
    t2tOutputFileName(None, None)
    
except Exception, e:
    msg =  "--------------- UNEXPECTED ERROR ---------------\n" + e.__str__()
    msg += "\n------------------------------------------------\n"
    msg += "\nThe application has been forcely closed!\n"
    bailOut(msg = msg, disconnect = tobii_connected)